grove/alpha/simon/simon.py

```diff
@@ -13,441 +13,325 @@
+        self.ancillas = self._qubits[self.n_qubits:]
+        self.classical_register = np.asarray(list(range(self.n_qubits + self.n_ancillas)))
+        self.unitary_function_mapping, _ = self._compute_unitary_oracle_matrix(bitstring_map)
+        self.simon_circuit = self._construct_simon_circuit()
+        self._dict_of_linearly_indep_bit_vectors = {}
+        self.mask = None
+        self.classical_register = np.asarray(list(range(self.n_qubits + self.n_ancillas)))
+
+    @staticmethod
+    def _compute_unitary_oracle_matrix(bitstring_map):
+        """
+        Computes the unitary matrix that encodes the orcale function for Simon's algorithm
+
+        :param bitstring_map: truth-table of the input bitstring map in dictionary format
+        :return: a dense matrix containing the permutation of the bit strings and a dictionary
+        containing the indices of the non-zero elements of the computed permutation matrix as
+        key-value-pairs
+        :rtype: Tuple[2darray, Dict[String, String]]
+        """
+        n_bits = len(list(bitstring_map.keys())[0])
+        ufunc = np.zeros(shape=(2 ** (2 * n_bits), 2 ** (2 * n_bits)))
+        dct = defaultdict(dict)
+        for b in range(2**n_bits):
+            pad_str = np.binary_repr(b, n_bits)
+            for k, v in bitstring_map.items():
+                dct[pad_str + k] = utils.bitwise_xor(pad_str, v) + k
+                i, j = int(pad_str+k, 2), int(utils.bitwise_xor(pad_str, v) + k, 2)
+                ufunc[i, j] = 1
+        return ufunc, dct
+
+    def find_mask(self, cxn, bitstring_map):
+        """
+        Runs Simon'mask_array algorithm to find the mask.
+
+        :param JobConnection cxn: the connection to the Rigetti cloud to run pyQuil programs
+        :param Dict[String, String] bitstring_map: a truth table describing the boolean function,
+        whose period is  to be found.
+
+        :return: Returns the mask (period) of the bitstring map or raises and Exception if the mask
+        cannot be found.
+        """
+        if not isinstance(bitstring_map, dict):
+            raise ValueError("Bitstring map needs to be a map from bitstring to bitstring")
+        self._init_attr(bitstring_map)
+
+        # create the samples of linearly independent bit-vectors
+        self._sample_independent_bit_vectors(cxn)
+        # try to invert the mask and check validity
+        self._invert_mask_equation()
+
+        if self._check_mask_correct():
+            return self.mask
+        else:
+            raise Exception("No valid mask found")
+
+    def _sample_independent_bit_vectors(self, cxn):
+        """This method samples :math:`n-1` linearly independent vectors using the Simon Circuit.
+        It attempts to put the sampled bitstring into a dictionary and only terminates once the
+        dictionary contains :math:`n-1` samples
+
+
+
+
+        :param cxn: Connection object to the Quantum Engine (QVM, QPU)
```

Comment:
I don't think the phrase `Quantum Engine` means anything- can we just say `to a QVM or QPU`?

Reply:
done